OMI33 requisition processor to ignore deleted requisitions

andreievg opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

This is a patch PR to main

Fixes https://github.com/openmsupply/open-msupply-internal/issues/33

# 👩🏻‍💻 What does this PR do? 

Filters out deletes from requisition processor.

This fixes sync freezing that was described in the issue (btw, this was validated with a custom build)

Sync was freezing because of deleted records being processed in requisition processor (only deleted requisitions were in changelog for processing)

https://github.com/openmsupply/open-msupply/blob/4e8bc15d60be9a5f5e630f0eb1c70fb443b496e2/server/service/src/processors/transfer/requisition/mod.rs#L113

Above line will continue, and not update cursors:

https://github.com/openmsupply/open-msupply/blob/4e8bc15d60be9a5f5e630f0eb1c70fb443b496e2/server/service/src/processors/transfer/requisition/mod.rs#L131-L136

Why does sync stop thought ? It looks like it's due to the following select, which I though would be fine on multi thread application, but for some reason the processor still blocks it (at least on the cloud server, look like the test passes ):

https://github.com/openmsupply/open-msupply/blob/4e8bc15d60be9a5f5e630f0eb1c70fb443b496e2/server/server/src/lib.rs#L280-L281

# 🧪 How has/should this change been tested? 

There is already a unit test for this, but can replicate this by creating, then deleting requisitions, then trying to sync (requisition deleting should be the last operation), not sure if this manual test would work on a normal mac, may need a special single threaded environment (since unit test does something similar where timeout running similar to synchroniser, but timeout is not blocked, unless test is set to one thread)

andreievg commented on package.json:
```diff
@@ -1,11 +1,11 @@
   "name": "open-msupply",
   "//": "Main version for the app, should be in semantic version format (any release candidate or test build should be separated by '-' i.e. 1.1.1-rc1 or 1.1.1-test",
-  "version": "1.1.15",
+  "version": "1.1.17",
```
Will build new version when this is merged

andreievg commented on server/service/src/processors/transfer/requisition/test.rs:
```diff
@@ -130,6 +132,77 @@ pub(crate) struct RequisitionTransferTester {
 }
 
+/// Deleted requisitions stuck forever
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
```
Test fails and freezes when using tokio::test (which is usually single threaded test)

clemens-msupply reviewed (approved):
Looks good!

I tested by switching back to the old code -> test fails. Otherwise, the test runs fine.

What about the error cases in the loop. If there is an error then we expect sync to hang, right?

Nit: you could put in a circuit breaker into the outer loop testing if you made progress since the last loop iteration... Then the rest of sync would keep working if there is a problem for the transferes (?)

andreievg commented:
> I tested by switching back to the old code -> test fails. Otherwise, the test runs fine.

Thank you.

> What about the error cases in the loop. If there is an error then we expect sync to hang, right?

I think it should be safe, error will be returned here and logged:

https://github.com/openmsupply/open-msupply/blob/7356712e454ac4ce45ac63e7fa312521eb39eec4/server/service/src/processors/mod.rs#L81

https://github.com/openmsupply/open-msupply/blob/7356712e454ac4ce45ac63e7fa312521eb39eec4/server/service/src/processors/mod.rs#L84

https://github.com/openmsupply/open-msupply/blob/7356712e454ac4ce45ac63e7fa312521eb39eec4/server/service/src/processors/mod.rs#L96-L98

But yeah, next time processor is called it will try to resume from that broken record, that would create a lot of new cases to think about (different data states ?), I added a comment: 9970d6e7c97e700cda4cb1f7c11cb790d441c9fe

> Nit: you could put in a circuit breaker into the outer loop testing if you made progress since the last loop iteration... Then the rest of sync would keep working if there is a problem for the transferes (?)

Like adding another update to RequisitionTransferProcessorCursor after the inner loop (using the last log cursor + 1) ? I added comment: 9970d6e7c97e700cda4cb1f7c11cb790d441c9fe

andreievg reviewed (commented):
Rever version change
